include/openmc/static_map.h

```diff
@@ -0,0 +1,251 @@
+#ifndef OPENMC_STATIC_MAP_H
+#define OPENMC_STATIC_MAP_H
+
+#include <algorithm> // for stable_sort, find_if, lower_bound
+#include <utility> // for pair
+
+#include "openmc/vector.h"
+
+namespace openmc {
+
+//==========================================================================
+// Integer hash function
+//==========================================================================
+
+//! Hash function for integers that simply returns the integer.
+//
+//! If the values in the hash table are consecutive starting from zero, this
+//! will produce no collisions. Otherwise, all bets are off.
+//
+//! TODO: explore the performance of different hash functions
+template<typename Key>
+struct integer_hash
+{
+  using value_type = Key;
+  using result_type = std::size_t;
+
+  result_type operator()(value_type value) const
+  {
+    return static_cast<result_type>(value);
+  }
+};
+
+//==========================================================================
+// Class declarations
+//==========================================================================
+
+//! Fixed-size map with on-device access.
+//
+//! This is a read-only hash table that supports on-device lookups. The table
+//! is constructed on the host and must be "finalized" before it can be used to
+//! search for items. Adapted from the \c Static_Device_Map class in Oak Ridge
+//! National Laboratory's SCALE code system.
```

Comment:
If any more info is known, it could be useful to give more detail on where the Static_Device_Map code originated (e.g., a link to a SCALE repo, or who shared the implementation, etc). It's possible we may have need more GPU-friendly data structures like this somewhere down the road so would be nice to have more info on where this came from for future reference.

Reply:
Good point -- Seth Johnson at ORNL passed along the Static_Device_Map.cuh/.i.cuh files, but I'm not sure they are publicly available anywhere. Quoting from slack:

"since it's upstream of our open source components and isn't a nuclear code, it should be open source. The hash tables we've used have gone through several iteations: tom made the first one with one vector per bucket, I made a static version of that one, steve ported that one to GPU".